### PR TITLE
Feature/restructure

### PR DESCRIPTION
**Pull request by gecBurton**
I have changed:
* I have tidied up the names and namespaces, i.e. `automatilib.automatilib.cola_auth` is now just `automatilib.cola`
* I have added a sample project, this is primarily so we can write unit tests against it
* I have added a lot of auth logic to the `conftest` so we can mimic how cola is working

I have not changed, the logic or basic code structure of the core code (much)

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -39,12 +39,18 @@ To make use of COLA
 -------------------
 
 - Add these into your `INSTALLED_APPS` settings:
-    - "automatilib.automatilib.core",
-    - "automatilib.automatilib.cola_auth",
+    - "core.apps.IdotAIConfig"
+    - "cola.apps.ColaAuthConfig"
```
Comment (gecBurton) [inline README.rst:43]: maybe should be    `"automatilib.apps.ColaAuthConfig"`

File: cola/backend.py [hunk 2]
```diff
@@ -20,15 +20,15 @@ def authenticate(self, request, username: Optional[str] = None, password: Option
...
         user.save()
         LOGGER.info(f"Set values: {user_response} - from COLA authentication")
         return user
 
-    def get_user(self, user_id):
-        try:
-            return User.objects.get(pk=user_id)
-        except User.DoesNotExist:
-            return None
+    # def get_user(self, user_id):
```
Comment (gecBurton) [inline cola/backend.py]: i dont think we need this, at least it isnt showing up in the test coverage?

File: cola/views.py [hunk 3]
```diff
@@ -16,19 +15,13 @@
 
 LOGGER = logging.getLogger(__name__)
 
-User = get_user_model()
 
+COLA_ISSUER = f"https://cognito-idp.{settings.AWS_REGION_NAME}.amazonaws.com/{settings.COLA_COGNITO_USER_POOL_ID}"
 
-def get_cola_cognito_user_pool_jwk() -> Optional[tuple[dict, str]]:
```
Comment (gecBurton) [inline cola/views.py:22]: i found a less aggressive way to mock this

File: cola/views.py [hunk 4]
```diff
@@ -50,13 +43,14 @@ def get(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
...
+
         redirect_url = settings.LOGIN_URL
-        response = redirect(reverse(redirect_url or "index"))
+        response = redirect(reverse(redirect_url))
         response.delete_cookie(settings.COLA_COOKIE_NAME)
         return response
 
 
-class IAIColaLogin(View):
+class ColaLogin(View):
```
Comment (gecBurton) [inline cola/views.py]: am open to what this should be called, but i think  we only need to say who we are once and not in ever class?

File: cola/views.py [hunk 5]
```diff
@@ -86,8 +80,8 @@ def handle_user_jwt_details(self, user: User, token_payload: dict) -> None:
 
     def get(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
         redirect_url = settings.LOGIN_REDIRECT_URL
-        if request.user.id:
-            return redirect(reverse(redirect_url or "index"))
```
Comment (gecBurton) [inline cola/views.py:90]: there is already a default fall back for this not being set

File: cola/views.py [hunk 6]
```diff
@@ -102,35 +96,38 @@ def get(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
...
-        if not jwk_response:
-            return HttpResponseServerError()  # TODO: Improve error handling throughout
-        else:
-            cola_cognito_user_pool_jwk, cola_issuer = jwk_response
+
+        response = requests.get(COLA_URL, timeout=5)
+        if response.status_code != 200:
+            LOGGER.error("failed to get expected response from COLA")
+            return HttpResponse("Unauthorized", status=401)
+        cola_cognito_user_pool_jwk = response.json()
```
Comment (gecBurton) [inline cola/views.py]: I returned this to how you had it originally @252afh, which i think is better, after i found a better way to mock it

File: cola/views.py [hunk 7]
```diff
@@ -102,35 +96,38 @@ def get(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
...
+
         header = jwt.get_unverified_header(regex_match.group(0))
         jwt_kid = header["kid"]
         public_key = next(key for key in cola_cognito_user_pool_jwk["keys"] if key["kid"] == jwt_kid)
 
+        token = regex_match.group(0)
+        token = ".".join(token.split(".")[:3])
```
Comment (gecBurton) [inline cola/views.py]:
```suggestion
```
i think we can come up with a regex expression that makes this step redundant

File: cola/views.py [hunk 8]
```diff
@@ -102,35 +96,38 @@ def get(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
...
                     "require_exp": True,
                     "require_iss": True,
                     "require_sub": True,
                 },
             )
 
-        except (ExpiredSignatureError, JWTClaimsError, JWTError, KeyError) as error:
-            LOGGER.error("cookie error:", type(error).__name__)
+        except (ExpiredSignatureError, JWTClaimsError, JWTError) as error:
+            LOGGER.error(f"cookie error: {error}")
```
Comment (gecBurton) [inline cola/views.py]: i was getting str conversion errors with out this

File: core/apps.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+from django.apps import AppConfig
+
+
+class IdotAIConfig(AppConfig):
```
Comment (gecBurton) [inline core/apps.py]:
```suggestion
class CoreConfig(AppConfig):
```

File: core/apps.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
...
+    name = "core"
+    label = "core"
+    verbose_name = "i.AI's COLA Authentication"
```
Comment (gecBurton) [inline core/apps.py]:
```suggestion
    verbose_name = "i.AI's core models and utities"
```

File: automatilib/cola/backend.py [hunk 11]
```diff
@@ -20,15 +20,15 @@ def authenticate(self, request, username: Optional[str] = None, password: Option
...
         user.save()
         LOGGER.info(f"Set values: {user_response} - from COLA authentication")
         return user
 
-    def get_user(self, user_id):
-        try:
-            return User.objects.get(pk=user_id)
-        except User.DoesNotExist:
-            return None
+    # def get_user(self, user_id):
```
Comment (252afh) [inline automatilib/cola/backend.py]: This is a required function for creating a new backend, even if it is never shown as being used

https://docs.djangoproject.com/en/4.2/topics/auth/customizing/#:~:text=An%20authentication%20backend%20is%20a,optional%20permission%20related%20authorization%20methods.

File: automatilib/cola/models.py
Comment (252afh) [inline automatilib/cola/models.py:1]: Not needed

File: automatilib/cola/settings.py
Comment (252afh) [inline automatilib/cola/settings.py:1]: Can move these here

```
INSTALLED_APPS = [
    "automatilib.cola",
]

AUTHENTICATION_BACKENDS = [
    "automatlib.cola.backend.AuthenticationBackend",
]
```

File: automatilib/cola/views.py [hunk 12]
```diff
@@ -74,7 +68,7 @@ def post_login(self):
         pass
 
     @abstractmethod
-    def handle_user_jwt_details(self, user: User, token_payload: dict) -> None:
+    def handle_user_jwt_details(self, user, token_payload: dict) -> None:
```
Comment (252afh) [inline automatilib/cola/views.py]: Why did we take away the `user` type, every user should inherit from djangos base user, or we can use django's method to get the instance of the user model in use

File: automatilib/cola/views.py [hunk 13]
```diff
@@ -86,8 +80,8 @@ def handle_user_jwt_details(self, user: User, token_payload: dict) -> None:
 
     def get(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
         redirect_url = settings.LOGIN_REDIRECT_URL
```
Comment (252afh) [inline automatilib/cola/views.py]: Are we assuming this has to be set and index might not be in place?

File: automatilib/cola/views.py [hunk 14]
```diff
@@ -102,50 +96,51 @@ def get(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
...
             return HttpResponse("Unauthorized", status=401)
-        jwk_response = get_cola_cognito_user_pool_jwk()
-        if not jwk_response:
-            return HttpResponseServerError()  # TODO: Improve error handling throughout
-        else:
-            cola_cognito_user_pool_jwk, cola_issuer = jwk_response
+
+        response = requests.get(COLA_URL, timeout=5)
+        if response.status_code != 200:
+            LOGGER.error("failed to get expected response from COLA")
```
Comment (252afh) [inline automatilib/cola/views.py]: Pedantic but capitalise the `f`

File: automatilib/cola/views.py [hunk 15]
```diff
@@ -16,19 +15,13 @@
...
+COLA_ISSUER = f"https://cognito-idp.{settings.AWS_REGION_NAME}.amazonaws.com/{settings.COLA_COGNITO_USER_POOL_ID}"
 
-def get_cola_cognito_user_pool_jwk() -> Optional[tuple[dict, str]]:
-    cola_issuer = f"https://cognito-idp.{settings.AWS_REGION_NAME}.amazonaws.com/{settings.COLA_COGNITO_USER_POOL_ID}"
-    response = requests.get(f"{cola_issuer}/.well-known/jwks.json", timeout=5)
-    if response.status_code != 200:
-        return None
-    cola_cognito_user_pool_jwk = response.json()
-    return cola_cognito_user_pool_jwk, cola_issuer
+COLA_URL = f"{COLA_ISSUER}/.well-known/jwks.json"
```
Comment (252afh) [inline automatilib/cola/views.py]: This should be called `COLA_JWK_URL` instead, `COLA_URL` to me would refer to their login catch url

File: automatilib/cola/views.py [hunk 16]
```diff
@@ -102,50 +96,51 @@ def get(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
...
 
         authenticated_user = {
             "email": payload["email"],
         }
 
         if user := authenticate(request=request, user_response=authenticated_user):
-            LOGGER.info("Attempting to log user in")
-            LOGGER.debug(user.__dict__)
-            user = authenticate(request=request, user_response=authenticated_user)
+            LOGGER.info(f"Attempting to log user {user.pk} in")
```
Comment (252afh) [inline automatilib/cola/views.py:140]: Missing authenticating the user here

File: automatilib/cola/urls.py [hunk 17]
```diff
@@ -0,0 +1,8 @@
+from django.urls import path
+
+from automatilib.cola import views
+
+url_patterns = [
+    path("login/", views.ColaLogin.as_view(), name="login"),
```
Comment (252afh) [inline automatilib/cola/urls.py]: This should be `post-login`, the `login` url is reserved for going out to COLA

File: automatilib/core/apps.py [hunk 18]
```diff
@@ -0,0 +1,7 @@
+from django.apps import AppConfig
+
+
+class IdotAIConfig(AppConfig):
+    name = "automatilib.core"
+    label = "core"
+    verbose_name = "i.AI's COLA Authentication"
```
Comment (252afh) [inline automatilib/core/apps.py]: verbose name is wrong

File: automatilib/core/migrations/__init__.py
Comment (252afh) [inline automatilib/core/migrations/__init__.py:1]: Don't need an init in migrations

File: automatilib/core/models.py [hunk 19]
```diff
@@ -27,3 +27,6 @@ class IDotAiUser(BaseUser, UUIDPrimaryKeyBase):
     def save(self, *args, **kwargs):
         self.email = self.email.lower()
         super().save(*args, **kwargs)
+
+    class Meta:
```
Comment (252afh) [inline automatilib/core/models.py:31]: If this is abstract then we can get rid of the migrations

File: automatilib/core/settings.py [hunk 20]
```diff
@@ -1,11 +1,11 @@
...
+    "automatilib.cola",
     "django_use_email_as_username.apps.DjangoUseEmailAsUsernameConfig",
 ]
 
 AUTHENTICATION_BACKENDS = [
-    "automatilib.automatilib.cola_auth.backend.AuthenticationBackend",
+    "automatlib.cola.backend.AuthenticationBackend",
 ]
 
 AUTH_USER_MODEL = "automatilib.core.User"
```
Comment (252afh) [inline automatilib/core/settings.py]: Can probably remove this if the user model is abstract

File: example_project/apps.py [hunk 21]
```diff
@@ -0,0 +1,7 @@
+from django.apps import AppConfig
+
+
+class ExampleProjectConfig(AppConfig):
+    name = "example_project"
+    label = "example_project"
+    verbose_name = "i.AI's COLA Authentication"
```
Comment (252afh) [inline example_project/apps.py]: verbose name is wrong

File: example_project/models.py [hunk 22]
```diff
@@ -0,0 +1,5 @@
+from automatilib.core.models import IDotAiUser, UUIDPrimaryKeyBase
+
+
+class MyUser(IDotAiUser, UUIDPrimaryKeyBase):
```
Comment (252afh) [inline example_project/models.py:4]: Do we need a migration for this user?

**Review by 252afh — CHANGES_REQUESTED** [review]

File: automatilib/cola/views.py (hunk 12 above)
Comment (gecBurton) [reply automatilib/cola/views.py]: ah good question, i forget exactly what the issue is, I think django was complaining that we were using a model that was registered to any app, probably we can use an abstract Model though.. i'll give it a go

File: automatilib/cola/views.py (hunk 13 above)
Comment (gecBurton) [reply automatilib/cola/views.py]: there is a default provided by the ModelBackend, i think it is `accounts/login`

File: automatilib/cola/views.py (hunk 12 above)
Comment (gecBurton) [reply automatilib/cola/views.py]: https://github.com/i-dot-ai/automatilib/pull/3#discussion_r1378537959 I think this is related

File: example_project/models.py (hunk 22 above)
Comment (gecBurton) [reply example_project/models.py:4]: ~apparently not as the tests pass without it, but probably worth adding one~

File: automatilib/cola/views.py (hunk 14 above)
Comment (gecBurton) [inline automatilib/cola/views.py]:
```suggestion
            LOGGER.error("Failed to get expected response from COLA")
```

File: automatilib/cola/views.py (hunk 16 above)
Comment (gecBurton) [reply automatilib/cola/views.py:140]: isnt this done on line 137?

File: automatilib/core/apps.py (hunk 18 above)
Comment (gecBurton) [inline automatilib/core/apps.py]:
```suggestion
    verbose_name = "i.AI's core models"
```

File: example_project/apps.py (hunk 21 above)
Comment (gecBurton) [inline example_project/apps.py]:
```suggestion
    verbose_name = "example application using COLA Authentication"
```

File: automatilib/cola/views.py (hunk 15 above)
Comment (gecBurton) [reply automatilib/cola/views.py]: [597dbe7](https://github.com/i-dot-ai/automatilib/pull/3/commits/597dbe74e264d93fc24d447d5537c78d135cb8b6)

File: automatilib/cola/urls.py (hunk 17 above)
Comment (gecBurton) [reply automatilib/cola/urls.py]: [6e0d166](https://github.com/i-dot-ai/automatilib/pull/3/commits/6e0d16647f6a42f3c13c0af558ea7bd0aa71f217)

File: automatilib/cola/models.py
Comment (gecBurton) [reply automatilib/cola/models.py:1]: [13d7c5f](https://github.com/i-dot-ai/automatilib/pull/3/commits/13d7c5f78e671231a8ea959d4b8e02bb249e0a86)

File: automatilib/cola/backend.py (hunk 11 above)
Comment (gecBurton) [reply automatilib/cola/backend.py]: [d58b30b](https://github.com/i-dot-ai/automatilib/pull/3/commits/d58b30bceeff850176332f1784af200f2df22346)

File: automatilib/cola/settings.py
Comment (gecBurton) [reply automatilib/cola/settings.py:1]: see [8793fb9](https://github.com/i-dot-ai/automatilib/pull/3/commits/8793fb9530152954a01da5642568abfeaec50fc0) 

I could remove stuff from the example_project/settings.py but adding it to cola/settings.py didnt make any difference

File: example_project/models.py (hunk 22 above)
Comment (gecBurton) [reply example_project/models.py:4]: added here [29bacb0](https://github.com/i-dot-ai/automatilib/pull/3/commits/29bacb061e44879b687bb97cc66207c85dbf51cc)

File: automatilib/cola/views.py (hunk 16 above)
Comment (252afh) [reply automatilib/cola/views.py:140]: So it is

File: automatilib/cola/settings.py
Comment (252afh) [reply automatilib/cola/settings.py:1]: Including it using string quotes makes this part redundant. We can import it into our django-apps settings file and do 

```
import settings from automatilib.core as AUTOMATILIB_SETTINGS 
import settings from automatilib.cola as COLA_SETTINGS 

INSTALLED_APPS = [
"other_stuff",
*COLA_SETTINGS.COLA_INSTALLED_APPS,
*AUTOMATILIB_SETTINGS.AUTOMATILIB_INSTALLED_APPS,
]
```

Or scrap the whole thing and let the user import using strings

File: automatilib/cola/views.py (hunk 13 above)
Comment (252afh) [reply automatilib/cola/views.py]: That relies on the user having all_auth in use. Plus this is the post-login url, so it's a just-in-case clause if for some weird reason they haven't set the url

File: automatilib/core/settings.py (hunk 20 above)
Comment (252afh) [reply automatilib/core/settings.py]: By `this` I mean `AUTH_USER_MODEL = "automatilib.core.User"` btw, highlighting isn't obvious in git for me

Comment (gecBurton) [reply automatilib/core/settings.py]: ah, i removed `AUTH_USER_MODEL = "automatilib.core.User` too

File: automatilib/cola/settings.py
Comment (gecBurton) [reply automatilib/cola/settings.py:1]: > Or scrap the whole thing and let the user import using strings

tbh i think i prefer this as its very clear?

Comment (gecBurton) [reply automatilib/cola/settings.py:1]: [ac32f6d](https://github.com/i-dot-ai/automatilib/pull/3/commits/ac32f6d629030e3dc909a7a60e864bf30722ce80)

> Or scrap the whole thing and let the user import using strings

i went for this option as it was clearer, happy to change though

File: automatilib/cola/views.py (hunk 13 above)
Comment (gecBurton) [reply automatilib/cola/views.py]: [de37f](https://github.com/i-dot-ai/automatilib/pull/3/commits/14de37fb4568088e40f778d5fe033389e2b167b4)

File: automatilib/cola/settings.py
Comment (252afh) [reply automatilib/cola/settings.py:1]: The core settings file can be deleted as well in that case :)

**Review by 252afh — APPROVED** [review]

**Pull request merged**